File: apprefurbished/src/main/java/com/androidcollider/easyfin/managers/accounts_info/AccountsInfoManager.java
package com.androidcollider.easyfin.managers.accounts_info;

import com.androidcollider.easyfin.models.Account;
import com.androidcollider.easyfin.repository.Repository;
import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;

import java.util.ArrayList;
import java.util.List;

import rx.Observable;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * @author Ihor Bilous
 */

public class AccountsInfoManager {

    private List<Account> accountList;
    private Repository repository;


    AccountsInfoManager(Repository repository) {
        this.repository = repository;
        accountList = new ArrayList<>();
        loadAccountList();
    }

    private List<String> getAccountNames() {
        return Stream.of(accountList)
                .map(Account::getName)
                .collect(Collectors.toList());
    }

    public boolean checkForAccountNameMatches(String name) {
        List<String> accountNames = getAccountNames();
        for (String account : accountNames) {
            if (account.equals(name)) return true;
        }
        return false;
    }

    public int getAccountsCount() {
        return accountList.size();
    }

    public Observable<Integer> getAccountsCountObservable() {
        return repository.getAllAccounts()
                .map(List::size)
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread());
    }

    private void loadAccountList() {
        repository.getAllAccounts()
                .subscribe(new Subscriber<List<Account>>() {

                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onNext(List<Account> accountList) {
                        AccountsInfoManager.this.accountList.clear();
                        AccountsInfoManager.this.accountList.addAll(accountList);
                    }
                });
    }
}

File: apprefurbished/src/main/java/com/androidcollider/easyfin/common/managers/accounts/accounts_info/AccountsInfoManager.kt
package com.androidcollider.easyfin.common.managers.accounts.accounts_info

import com.androidcollider.easyfin.common.models.Account
import com.androidcollider.easyfin.common.repository.Repository
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.schedulers.Schedulers

/**
 * @author Ihor Bilous
 */
class AccountsInfoManager internal constructor(private val repository: Repository) {
    private val accountList: MutableList<Account>

    init {
        accountList = ArrayList()
        loadAccountList()
    }

    private val accountNames: List<String>
        get() = accountList.map(Account::name)

    fun checkForAccountNameMatches(name: String): Boolean {
        val accountNames = accountNames
        for (account in accountNames) {
            if (account == name) return true
        }
        return false
    }

    val accountsCountObservable: Single<Int>
        get() = repository.allAccounts!!.map { obj: List<Account> -> obj.size }
            .subscribeOn(Schedulers.computation())
            .observeOn(AndroidSchedulers.mainThread())

    private fun loadAccountList() {
        repository.allAccounts?.subscribe(
            { accounts: List<Account> ->
                accountList.clear()
                accountList.addAll(accounts)
            }) { obj: Throwable -> obj.printStackTrace() }
    }
}